src/emit.rs

```diff
@@ -0,0 +1,170 @@
+// Copyright 2018-2024 the Deno authors. All rights reserved. MIT license.
+
+use std::rc::Rc;
+
+use anyhow::Result;
+use base64::Engine;
+use swc_common::comments::SingleThreadedComments;
+use swc_common::SourceMap;
+
+use crate::swc::ast::Program;
+use crate::swc::codegen::text_writer::JsWriter;
+use crate::swc::codegen::Node;
+use crate::swc::common::FileName;
+use crate::ModuleSpecifier;
+
+#[derive(Default, Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub enum SourceMapOption {
+  /// Source map should be inlined into the source (default)
+  #[default]
+  Inline,
+  /// Source map should be generated as a separate file.
+  Separate,
+  /// Source map should not be generated at all.
+  None,
+}
+
+#[derive(Debug, Clone, Copy, Hash)]
+pub struct EmitOptions {
+  /// How and if source maps should be generated.
+  pub source_map: SourceMapOption,
+  /// Whether to inline the source contents in the source map. Defaults to `true`.
+  pub inline_sources: bool,
+  /// Whether to keep comments in the output. Defaults to `false`.
+  pub keep_comments: bool,
+}
+
+impl Default for EmitOptions {
+  fn default() -> Self {
+    EmitOptions {
+      source_map: SourceMapOption::default(),
+      inline_sources: true,
+      keep_comments: false,
+    }
+  }
+}
+
+/// An Emitter is used to emit a source file based on the emit options.
+pub struct Emitter {
```

Comment:
Not sure about this `Emitter` holding state like this and being used for transpiling before being used for emitting and then having a public `source_map` property like that. Maybe `emit` should just be a function and then the calller can continue to construct a `SourceMap`?

I was going to suggest we make a wrapper around `SourceMap` for these single file emit scenarios, but I actually was just diving into the swc code just now and realized that we can completely get rid of using `SourceMap` with swc for these cases by implementing the `SourceMapper` trait on an object, which should make emitting slightly faster.

Reply:
Yes that's why I did this (the JSX transform). I tried what you suggested first but it didn't turn out so great.

Reply:
I think we should still change `Emitter` to be an `emit` function instead though because this `Emitter` has two responsibilities atm (hence the public `sourcemap` property) and is required to be constructed before transpiling. We can come up with something for reducing the code for `SourceMap` in the future.